### PR TITLE
Pick body parser

### PR DESCRIPTION
**Pull request by supertong**
This should allow the middleware to pick the correct body-parser base on the definition in schema and config.

### CODE REVIEW

**Comment by supertong** [comment]
@lukiano

File: src/index.ts [hunk 1]
```diff
@@ -92,7 +73,50 @@ export function oas(cfg: Partial<Config>): koa.Middleware {
...
+    }
+
+    if (!config.validatePaths.some(path => ctx.path.startsWith(path))) {
+      // Skip validation if no path matches
+      return next();
+    }
+
+    const middlewares: Array<koa.Middleware> = [];
+    const requestContentTypes = compiled.getDefinedRequestBodyContentType(ctx.path, ctx.request.method);
+    // const matchedContentType = reqContentTypeIs(ctx.req, requestContentTypes);
```
Comment (lukiano) [inline src/index.ts]: do we need this line?

File: __tests__/helpers/createContext.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  // Some functions we call in the implementations will perform checks for `req.encrypted`, which delegates to the socket.
+  // MockRequest doesn't set a fake socket itself, so we create one here.
+  mockedReq.socket = new Duplex() as any;
+  Object.defineProperty(mockedReq.socket, 'encrypted', {
+    writable: false,
+    value: false,
+  });
+
+  if (req.body) {
+    process.nextTick(() => {
```
Comment (lukiano) [inline __tests__/helpers/createContext.ts]: I can imagine why this is needed, but perhaps we could add some explicit comment?

File: src/index.ts (hunk 1 above)
Comment (supertong) [reply src/index.ts]: Ahh, nope. I forgot to clean this up.

**Comment by supertong** [comment]
From my understanding, minor bump for version under 1 is a major bump.

File: __tests__/helpers/createContext.ts (hunk 2 above)
Comment (supertong) [reply __tests__/helpers/createContext.ts]: Added.

**Comment by lucaslago** [comment]
LGTM, as discussed offline: might be worth it adding a test case checking we don't consume the stream when there's no content-type or matching content-type parser since that caused us problem in the past

File: src/index.ts [hunk 3]
```diff
@@ -92,7 +73,49 @@ export function oas(cfg: Partial<Config>): koa.Middleware {
...
+        middlewares.push(config.requestBodyHandler[`${parts[0]}/*`]);  // For a match like `application/*`
+      } else if (config.requestBodyHandler[`*/${parts[1]}`]) {
+        middlewares.push(config.requestBodyHandler[`*/${parts[1]}`]);  // For a match like `*/json`
+      } else if (config.requestBodyHandler['*/*']) {
+        middlewares.push(config.requestBodyHandler['*/*']);  // For a global type match defined as `*/*`
+      }
+    }
+
+    middlewares.push(validatorMW);
+    await compose(middlewares).call(this, ctx, next);
```
Comment (supertong) [inline src/index.ts]: It's recommended to use `.call` from the official example.
https://github.com/koajs/examples/blob/master/conditional-middleware/app.js#L16

File: __tests__/helpers/createContext.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  Object.defineProperty(mockedReq.socket, 'encrypted', {
+    writable: false,
+    value: false,
+  });
+
+  if (req.body) {
+    /**
+     * .send() method will emit the data event straight away, however,
+     * we want it to be fired only after we have attached all the event
+     * listerners.
```
Comment (lukiano) [inline __tests__/helpers/createContext.ts]: typo, but not a big deal

**Pull request merged**